### PR TITLE
[Taken]Fix issue #783 & #764

### PR DESCRIPTION
**Pull request by jmecn**
Fix issue #783 & #764

### CODE REVIEW

File: jme3-core/src/main/java/com/jme3/effect/shapes/EmitterSphereShape.java [hunk 1]
```diff
@@ -95,11 +95,10 @@ public void cloneFields( Cloner cloner, Object original ) {
...
     public void getRandomPoint(Vector3f store) {
-        do {
-            store.x = (FastMath.nextRandomFloat() * 2f - 1f) * radius;
-            store.y = (FastMath.nextRandomFloat() * 2f - 1f) * radius;
-            store.z = (FastMath.nextRandomFloat() * 2f - 1f) * radius;
-        } while (store.distance(center) > radius);
+        store.x = (FastMath.nextRandomFloat() * 2f - 1f) * radius;
+        store.y = (FastMath.nextRandomFloat() * 2f - 1f) * radius;
+        store.z = (FastMath.nextRandomFloat() * 2f - 1f) * radius;
+        store.addLocal(center);
```
Comment (shamanDevel) [inline jme3-core/src/main/java/com/jme3/effect/shapes/EmitterSphereShape.java]: Now it creates a random point within a box of the desired side length (=radius) around the center.
The Do-while loop is still necessary to skip the samples outside the sphere

**Comment by TehLeo** [comment]
Yes, just as shaman pointed out, it got to pick an uniform point on a sphere. The current method, picks a random point in a box, repeatedly until the point is in the sphere. It forgets to add the center.

Here are two methods:

```
 do {
    store.x = (FastMath.nextRandomFloat() * 2f - 1f);
    store.y = (FastMath.nextRandomFloat() * 2f - 1f);
    store.z = (FastMath.nextRandomFloat() * 2f - 1f);
} while (store.dot(store) > 1);
store.multLocal(radius);
store.addLocal(center);
```

Or a method without a while loop. (ref: https://math.stackexchange.com/questions/87230/picking-random-points-in-the-volume-of-sphere-with-uniform-probability  )
```
float l = FastMath.pow(FastMath.nextRandomFloat(), 1f/3f);
float u = FastMath.nextRandomFloat()*2f-1f;
float o = FastMath.nextRandomFloat()*FastMath.TWO_PI;
		
store.z = radius*l*u;
u = FastMath.sqrt(1f-u*u);
		
store.x = radius*l*u*FastMath.cos(o);
store.y = radius*l*u*FastMath.sin(o);
```

File: jme3-core/src/main/java/com/jme3/effect/shapes/EmitterSphereShape.java (hunk 1 above)
Comment (Nehon) [reply jme3-core/src/main/java/com/jme3/effect/shapes/EmitterSphereShape.java]: mhhh... wouldn't it be simpler to do;
```
     store.x = FastMath.nextRandomFloat();
     store.y = FastMath.nextRandomFloat();
     store.z = FastMath.nextRandomFloat();
     store.normalizeLocal(); // <- random direction
     store.multLocal(radius); // <- mult by the radius to be inside the sphere radius
     store.addLocal(center); // <- relocal aroudn the center.
```
No loop needed and the normalize ensure that the point is inside the sphere.

Comment (TehLeo) [reply jme3-core/src/main/java/com/jme3/effect/shapes/EmitterSphereShape.java]: It would be simpler, but would not be uniformly distributed.

Comment (Nehon) [reply jme3-core/src/main/java/com/jme3/effect/shapes/EmitterSphereShape.java]: Uh yeah first the x2-1 should stay to allow negative values. But then just take à random length between 0 and radius, and you break the uniformity.

Comment (shamanDevel) [reply jme3-core/src/main/java/com/jme3/effect/shapes/EmitterSphereShape.java]: I might be wrong, but I think there is needs to more probability mass on the points with bigger radius than on the inner points. So if you normalize and then scale with a value between 0 and radius, that distribution must not be uniform, but something like u'=sqrt(1-u^2), as TehLeo suggested.

**Comment by jmecn** [comment]
I see. Next time I will make separated PR.
Just modified the code, please check it.

**Comment by TehLeo** [comment]
I've checked your changes. It creates random points within a sphere volume.
However, the points are not uniformly distributed. 

See the two methods I posted in the comment above.

**Comment by jmecn** [comment]
ok, let me see..

**Comment by TehLeo** [comment]
Almost there. Except the line:

`store.multLocal(FastMath.nextRandomFloat() * radius);`

I have to say you are really trying to change the code I provided, which is fine, but try to understand the problem as well.

The random point was already selected uniformly within unit sphere volume within the while loop.
What is left to do is multiply it by radius and add the center.

Thus change that line back to.
`store.multLocal(radius);`

**Comment by jmecn** [comment]
Oh my.. I finally understand that line now.

**Comment by Nehon** [comment]
@TehLeo good for you?

**Comment by TehLeo** [comment]
Yes, all is fine with #764 .

**Comment by Nehon** [comment]
@jmecn Could you please rebase -i your PR so that it's only 2 commits one for each issue?
Do you know how to use rebase -i?

**Comment by jmecn** [comment]
@Nehon never use rebase before. But I think I can have a try.

**Comment by Nehon** [comment]
Here is a nice an thorough explanation https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

**Comment by jmecn** [comment]
@Nehon sir I have conbined the commits, please check my code.

**Comment by Nehon** [comment]
Perfect, nice work.
I'll merge this PR once the build pass

**Comment by jmecn** [comment]
Why this pr build failed? any reason?

**Comment by empirephoenix** [comment]
Seems like an error with the buildserver (all buildlogs are null/empty) , I suggest to just commit a extra space or similar somewhere, so that it is triggered again.

**Comment by Nehon** [comment]
No don't commit again I'll relaunch the build

**Pull request merged**